File: python/sum-of-multiples/sum_of_multiples.py
def sum_of_multiples(limit, factors):
    return sum(all_multiples(limit, factors))


def all_multiples(limit, factors):
    multiples = set()
    for factor in factors:
        multiples = multiples.union(get_multiples(limit, factor))
    return multiples

def get_multiples(limit, factor):
    if factor == 0:
        return []

    multiples = set()
    for i in range(0, limit):
        if i % factor == 0:
            multiples.add(i)
    return multiples

File: go/sum-of-multiples/sum_of_multiples.go
package summultiples

import "fmt"

func SumMultiples(limit int, divisors ...int) (sum int) {
	multiples := getMultiples(limit, divisors)
	for multiple := range multiples {
		sum += multiple
	}
	return sum
}

func getMultiples(limit int, divisors []int) (multiples map[int]bool) {
	multiples = map[int]bool{}
	for _, divisor := range divisors {
		multiplesForDivisor := getMultiplesForDivisor(limit, divisor)
		for _, multiple := range multiplesForDivisor {
			multiples[multiple] = true
		}
	}
	return multiples
}

func getMultiplesForDivisor(limit int, divisor int) (multiples []int) {
	if divisor == 0 {
		return multiples
	}
	multiplyer := 1
	candidate := divisor * multiplyer
	for candidate < limit {
		fmt.Printf("candidate %d, multiplyer %d, limit %d\n", candidate, multiplyer, limit)
		multiples = append(multiples, candidate)
		multiplyer += 1
		candidate = divisor * multiplyer
	}
	fmt.Printf("returning multiples %v\n", multiples)
	return multiples
}
